### ISSUE TITLE
Plan for IPC revival

### ISSUE THREAD

**Issue by rphmeier** [paritytech/parity#4591]
- Seamless IPC-codegen:
    - [ ] Remove "BinaryEncodable" trait and derive
    - [ ] Write `serde::Serializer` and `serde::Deserializer` implementations in `parity_ipc` crate.
    - [ ] Codegen in-place: no more `include!`: kill all "types" modules.
    - [ ] Explicitly fail on trait inheritance.
- Modularization:
    - [ ] Allow multiple IPC communication channels over a single socket
    - [ ] Make a distinction between stable IPC traits and internal utility traits.

...will add more to this list as they're thought of.

Other concerns:
  - Integration with futures-rs instead of blocking on every call?
  - Non-awful error messages in codegen?

**Comment by rphmeier** [comment]
@NikVolf thoughts?

**Comment by NikVolf** [comment]
@rphmeier
Well, first of all it should be run on the CI (@arkpar have turned it off last time). See no point at maintaining code that anybody can break without even noticing.

- IPC does not block on void functions. Futures can be cool but you can't preserve signatures with them. Or you should require that IPC is valid only for functions that return futures.
- I don't think there is need in any serde code there
  `bincode` that claimed to provide serde implementation of binary serialization was at least sub-optimal last time i checked it. We will also never be able to optimize for zero-copy messaging with serde.
- With procedural macros, yes, I can get rid of the `includes!` and some other ugliness

**Comment by rphmeier** [comment]
>IPC does not block on void functions. Futures can be cool but you can't preserve signatures with them. Or you should require that IPC is valid only for functions that return futures.

In most cases, the only async portion of the computation would be the IPC transport itself. The main use-case which I envision would be turning `Vec<T>` return types into a `BoxStream<T, IpcError>` so we can begin processing them even before the whole list is written to the socket.

It's a bit of a stretch goal but actually should increase performance quite a bit; for example, a simple `join` on various IPC futures -> a synchronous combination operation would play really nicely with our futures-based RPC. In the sync implementation, we could start writing a response packet as items are returned from the socket (actually, we could take it one step further and write them to the socket as we stream them from the DB) to seriously lower latency.

> I don't think there is need in any serde code there
We don't really _need_ it, but having it work with existing serialization frameworks would be a huge plus. Just slap a `#[derive(Serialize, Deserialize)]` on your struct and it's IPC ready! I think the overhead from not handling ZSTs perfectly (and last I checked the `BinaryEncodable` implementation for any non-`()` ZSTs actually was broken) would be dwarfed by the performance gains of doing async I/O properly.

> Well, first of all it should be run on the CI
IMO should enable it on the CI as soon as soon as the biggest annoyances are taken care of so it doesn't get in the way while developing.

**Comment by rphmeier** [comment]
(The "subtly broken" ZST handling can be seen here: https://github.com/ethcore/parity/blob/8404edb656693270ec1cd3956d204f625d28ec7c/ipc/rpc/src/binary.rs#L136)

If you have another ZST, like `struct Foo;` which you implement with `size(&self) -> 0` a `Result<(), Foo>` will always come out as `Err`.

**Comment by NikVolf** [comment]
@rphmeier
Again, about futures - you can't preserve signatures of the original trait and use futures. We can make codegen that converts all returning `Vec<T>` into anything you want, including `Stream<T>` but it will not hold the requirement of identical signatures so the application can run both with and without IPC.

ZST - I'm aware of that serialization problem, that's quite easy to fix, we just had no interfaces that could run into that problem, and i didn't bother to fix it.

And btw, it's just slap a `#[binary]` now on your struct and it's IPC ready! There is no 'existing serialization  framework' for serde besides `bincode` afaik. At the time i was considering it, it was pretty awful. Maybe now it has come better, I'll look into it.

**Comment by rphmeier** [comment]
@maciejhirsz mentioned that `bincode` is being re-written with `serde` and should be faster now. One other alternative might be cap'n proto, but we'll have to see if it fits our requirements.

We discussed earlier and thought that requiring `Future` and `Stream` return types actually does make sense for IPC-traits, and should be feasible to implement. Fancier stuff like auto-streaming collections probably won't be.

**Comment by NikVolf** [comment]
https://github.com/NikVolf/ipc-show/tree/098289a8a761ee167f766e069479e11652d242b2

first working prototype

but
it will require event loop running on the client side even without ipc

see test `conn`:

```rust
   #[test]
    fn conn() {
        run_server("0.0.0.0:22010");
        thread::sleep(::std::time::Duration::from_millis(100));
    
        let mut core = Core::new().unwrap();

        println!("Connecting...");
        let client = TcpClient::new(IpcProto);
        let handle = core.handle();
        let client_service = core.run(client.connect(&"127.0.0.1:22010".parse().unwrap(), &handle)).unwrap();

        println!("Invoking...");
        let client = IpcClient::new(0, client_service);
        let block = core.run(client.load(5)).unwrap();

        assert_eq!(block.number, 5);
    }
```

particulary line `let block = core.run(client.load(5)).unwrap()` (where `core` is event loop)

i'm going to get rid of it introducing custom embedded event loop

**Comment by 5chdn** [comment]
Is anyone actively working on this?

**Comment by NikVolf** [comment]
Nope, there are some concepts, but nobody has agreed on the best way to reimplement

**Comment by rphmeier** [comment]
The language has definitely developed so we can do this better now. The most important thing to me is that we do it in such a way that connecting different "channels" of communication to sub-binaries is not any more difficult than doing it without IPC. 

Then we won't end up feeling pressure to put functionality in places it doesn't belong. But I still don't think it's a huge priority.

**Comment by debris** [comment]
closed in #6795
